Content.Shared/Xenoarchaeology/Artifact/SharedXenoArtifactSystem.Node.cs

```diff
@@ -0,0 +1,314 @@
+        UpdateNodeResearchValue((ent, ent.Comp));
+        Dirty(ent);
+    }
+
+    public Entity<XenoArtifactNodeComponent> CreateNode(Entity<XenoArtifactComponent> ent, ProtoId<XenoArchTriggerPrototype> trigger, int depth = 0)
+    {
+        return CreateNode(ent, PrototypeManager.Index(trigger), depth);
+    }
+
+    public Entity<XenoArtifactNodeComponent> CreateNode(Entity<XenoArtifactComponent> ent, XenoArchTriggerPrototype trigger, int depth = 0)
+    {
+        var proto = PrototypeManager.Index(ent.Comp.EffectWeights).Pick(RobustRandom);
+
+        AddNode((ent, ent), proto, out var nodeEnt, dirty: false);
+        DebugTools.Assert(nodeEnt.HasValue, "Failed to create node on artifact.");
+
+        nodeEnt.Value.Comp.Depth = depth;
+        nodeEnt.Value.Comp.TriggerTip = trigger.Tip;
+        EntityManager.AddComponents(nodeEnt.Value, trigger.Components);
+
+        Dirty(nodeEnt.Value);
+        return nodeEnt.Value;
+    }
+
+    public bool HasUnlockedPredecessor(Entity<XenoArtifactComponent> ent, EntityUid node)
+    {
+        var predecessors = GetDirectPredecessorNodes((ent, ent), node);
+        if (predecessors.Count != 0 && predecessors.All(p => p.Comp.Locked))
+            return false;
+        return true;
+    }
+
+    public bool IsNodeActive(Entity<XenoArtifactComponent> ent, EntityUid node)
+    {
+        return ent.Comp.CachedActiveNodes.Contains(GetNetEntity(node));
+    }
+
+    public List<Entity<XenoArtifactNodeComponent>> GetActiveNodes(Entity<XenoArtifactComponent> ent)
+    {
+        return ent.Comp.CachedActiveNodes.Select(n => _nodeQuery.Get(GetEntity(n))).ToList();
+    }
+
+    public int GetResearchValue(Entity<XenoArtifactNodeComponent> ent)
+    {
+        if (ent.Comp.Locked)
+            return 0;
+
+        return ent.Comp.ResearchValue - ent.Comp.ConsumedResearchValue;
+    }
+
+    public void SetConsumedResearchValue(Entity<XenoArtifactNodeComponent> ent, int value)
+    {
+        ent.Comp.ConsumedResearchValue = value;
+        Dirty(ent);
+    }
+
+    public string GetNodeId(EntityUid uid)
+    {
+        return (CompOrNull<NameIdentifierComponent>(uid)?.Identifier ?? 0).ToString("D3");
+    }
+
+    public List<List<Entity<XenoArtifactNodeComponent>>> GetSegments(Entity<XenoArtifactComponent> ent)
+    {
+        var output = new List<List<Entity<XenoArtifactNodeComponent>>>();
+
+        foreach (var segment in ent.Comp.CachedSegments)
+        {
+            var outSegment = new List<Entity<XenoArtifactNodeComponent>>();
+            foreach (var netNode in segment)
+            {
+                var node = GetEntity(netNode);
+                outSegment.Add((node, XenoArtifactNode(node)));
+            }
+
+            output.Add(outSegment);
+        }
+
+        return output;
+    }
+
+    public Dictionary<int, List<Entity<XenoArtifactNodeComponent>>> GetDepthOrderedNodes(IEnumerable<Entity<XenoArtifactNodeComponent>> nodes)
+    {
```

Comment:
same here. Code looks like its place in separate domain object that can restore its state..